### PR TITLE
Module 9

### PR DESCRIPTION
**Pull request by gmoraita**
Included Step 1-4. 
NOTE: Hesitated applying step 6 given the comment in the instructions that may break things.

### CODE REVIEW

File: todo_app/data/boardelements.py [hunk 1]
```diff
@@ -1,31 +1,27 @@
...
-        self.name = board_dict.get(self.board_name,'') 
-        self.statuses = board_dict.get(self.board_statuses,{})  
-
 class BoardStatus():
-    boardstatus_id = ''
-    boardstatus_name = ''
-    boardstatus_pos = ''
+    _TODO = 'To Do'
+    _DOING = 'Doing'
+    _DONE = 'Done'
```
Comment (hugh-emerson) [inline todo_app/data/boardelements.py]: The underscore prefix is a Python convention to indicate that this should not be accessed from outside the class. Which is not what you want

File: todo_app/data/boardelements.py [hunk 2]
```diff
@@ -1,31 +1,27 @@
...
+    _statuses = [_TODO, _DOING, _DONE]
+    boardstatus_id = 'id'
+    boardstatus_name = 'name'
+    boardstatus_pos = 'pos'
     def __init__(self, status_dict: dict = {}):
         self.id = status_dict.get(self.boardstatus_id,'')
         self.name = status_dict.get(self.boardstatus_name,'')
         self.pos = status_dict.get(self.boardstatus_pos,'')
```
Comment (hugh-emerson) [inline todo_app/data/boardelements.py]: You can delete `boardstatus_id`, `boardstatus_name`, `boardstatus_pos` and the __init__ function right? In fact you don't even need a class, you just need the constants for TODO, DOING, DONE and a list grouping them up

File: todo_app/templates/table_macro.html [hunk 3]
```diff
@@ -19,34 +19,32 @@
...
               {# If first status then 0 (ToDo), if last status then 2 (Done) for any intermediate status then 1 (In Progress) #}
               
-              {% set statusesId_list = statuses.keys()|list%}
-              {% if statusesId_list[0] == item.status %} {%set pos=0%}
-              {% elif statusesId_list.index(item.status) == statusesId_list|length-1 %} {%set pos=2%}
+              
+              {% if item.status == statuses.0 %} {%set pos=0%}
+              {% elif item.status == statuses|last %} {%set pos=2%}
               {% else %} {%set pos=1%}
               {% endif %}
```
Comment (hugh-emerson) [inline todo_app/templates/table_macro.html]: If this is so unintuitive that you felt a code comment was required, why not write it differently.
E.g.

```
              {% if item.status == statuses.0 %} {%set status_icon='/static/status_0.png'%}
              {% elif item.status == statuses|last %} {%set status_icon='/static/status_2.png'%}
              {% else %} {%set set status_icon='/static/status_1.png'%}
              {% endif %}
```

You could actually make the png names more helpful too.

File: todo_app/tests/test_requests.py [hunk 4]
```diff
@@ -1,49 +1,98 @@
...
-@patch('requests.request')
-def test_index_page(mock_get_requests, client):
-    mock_get_requests.side_effect = mock_responses
+def test_show_tasks(app_client):
+    tasks = get_db_collection() 
+    tasks.insert_many(initial_mock_tasks)
+    
+    response = app_client.get('/')
+    assert response.status_code == 200 
+    assert tasks.find({Item.item_title: "Task 1"}) is not None
```
Comment (hugh-emerson) [inline todo_app/tests/test_requests.py:69]: This assertion is unrelated to the index page loading correctly. You're just testing your hardcoded test data

File: todo_app/tests/test_requests.py [hunk 5]
```diff
@@ -1,49 +1,98 @@
...
 
-    response = client.get('/')
+def test_add_task(app_client):
+    tasks = get_db_collection() 
+
+    response = app_client.post('/add', data=dict(new_task))
     assert response.status_code == 200 
+    assert tasks.find({Item.item_title: "Task 3"}) is not None
```
Comment (hugh-emerson) [inline todo_app/tests/test_requests.py]: Using `new_task['Item.item_title]` rather than just `"Task 3"` would make the test a bit clearer (/more robust)

Also there's no real reason for `new_task` to be defined at the top of the file rather than inside this one test where it's used.

File: todo_app/tests/test_requests.py [hunk 6]
```diff
@@ -1,49 +1,98 @@
...
+    tasks = get_db_collection() 
+    inserted_id = tasks.insert_one(task_to_be_deleted).inserted_id
     
-    html_string = str(response.data)
-    assert 'Tempting to say' in html_string
-    assert html_string.count('status_0') == 4 # 4 in todo
-    assert html_string.count('status_1') == 3 # 3 doing
-    assert html_string.count('status_2') == 5 # 5 done
\ No newline at end of file
+    assert tasks.find_one({Item.item_id: inserted_id}) is not None
```
Comment (hugh-emerson) [inline todo_app/tests/test_requests.py:86]: This is assertion is just testing that mongomock works correctly. You can delete this. Similarly for test_move_task

**Review by hugh-emerson — APPROVED** [review]
Not sure what you mean by "step 6", I assume  you meant part 5 - the IP stretch goal. It's not really that risky, but also not that important. (We'll stop using Heroku in module 11)

But anyway, this looks good 👍 I've left some suggestions but merge when you're ready. See my other comments for some suggestions.

File: todo_app/data/boardelements.py (hunk 1 above)
Comment (gmoraita) [reply todo_app/data/boardelements.py]: Done

File: todo_app/data/boardelements.py (hunk 2 above)
Comment (gmoraita) [reply todo_app/data/boardelements.py]: Done - added all under Item

File: todo_app/templates/table_macro.html (hunk 3 above)
Comment (gmoraita) [reply todo_app/templates/table_macro.html]: Very good suggestion. Changed and also made changes to the png name

File: todo_app/tests/test_requests.py (hunk 5 above)
Comment (gmoraita) [reply todo_app/tests/test_requests.py]: Done - good point

File: todo_app/tests/test_requests.py (hunk 4 above)
Comment (gmoraita) [reply todo_app/tests/test_requests.py:69]: Removed

File: todo_app/tests/test_requests.py (hunk 6 above)
Comment (gmoraita) [reply todo_app/tests/test_requests.py:86]: I was thinking of keeping as a proof that the function works as expected
